File: dag_py/src/wising_dag/core/india_switches.py
"""Layer 1 India's per-head income switches, enforced before the graph runs —
mirrors prototypes/graph-pilot/india-switches.js exactly (see its header):
a head whose switch is explicitly False carries no amounts, at the top level
and in every quarter."""
from __future__ import annotations


def _clean_slice(slice_):
    if not isinstance(slice_, dict):
        return slice_
    out = dict(slice_)
    di = out.get("domestic_income")
    if isinstance(di, dict):
        di = dict(di)
        out["domestic_income"] = di
        sal = di.get("salary")
        if isinstance(sal, dict) and sal.get("has_salary_income") is False:
            di["salary"] = {"has_salary_income": False}
        hp = di.get("house_property")
        if isinstance(hp, dict) and hp.get("has_house_property_income") is False:
            di["house_property"] = {"has_house_property_income": False, "properties": []}
        bi = di.get("business_income")
        if isinstance(bi, dict) and bi.get("has_business_or_fo_income") is False:
            di["business_income"] = {"has_business_or_fo_income": False, "entity_type": bi.get("entity_type")}
        if di.get("has_agricultural_income") is False:
            di["agricultural_income_inr"] = 0
    os_ = out.get("other_sources")
    if isinstance(os_, dict) and os_.get("has_other_sources_income") is False:
        out["other_sources"] = {"has_other_sources_income": False}
    return out


def apply_india_income_switches(india):
    if not isinstance(india, dict):
        return india
    out = _clean_slice(india)
    quarters = india.get("quarters")
    if isinstance(quarters, dict):
        out["quarters"] = {q: _clean_slice(v) for q, v in quarters.items()}
    return out

File: prototypes/graph-pilot/india-switches.js
"use strict";
/* ============================================================================
 * Layer 1 India's per-head "does this income apply?" switches, enforced
 * before the graph runs (26 Sep 2026).
 *
 * Switching an income head OFF in layer1_india.html only hides its section:
 * the amounts already typed (or, before the hydration fix, copied in from
 * Layer 1 US) stay in state. No engine node checked the switches, so a
 * switched-off salary was still taxed and still drove findings — the
 * visible symptom was a salary conflict on a client whose Layer 1 India
 * showed no salary at all.
 *
 * applyIndiaIncomeSwitches(india) returns a copy where every head whose
 * switch is explicitly false carries no amounts, at the top level and in
 * every quarter. Only an explicit false counts — a missing switch is left
 * alone, so hand-authored data without switches is unchanged. Called from
 * every entry point that builds the graph ctx (analyze.js, monitor-next's
 * dag-adapter.js); mirrored in dag_py's core/india_switches.py.
 * ==========================================================================*/
function cleanSlice(slice) {
  if (!slice || typeof slice !== "object") return slice;
  var out = Object.assign({}, slice);
  var di = out.domestic_income;
  if (di && typeof di === "object") {
    di = out.domestic_income = Object.assign({}, di);
    if (di.salary && di.salary.has_salary_income === false) di.salary = { has_salary_income: false };
    if (di.house_property && di.house_property.has_house_property_income === false) di.house_property = { has_house_property_income: false, properties: [] };
    if (di.business_income && di.business_income.has_business_or_fo_income === false) {
      di.business_income = { has_business_or_fo_income: false, entity_type: di.business_income.entity_type };
    }
    if (di.has_agricultural_income === false) di.agricultural_income_inr = 0;
  }
  if (out.other_sources && out.other_sources.has_other_sources_income === false) out.other_sources = { has_other_sources_income: false };
  return out;
}

function applyIndiaIncomeSwitches(india) {
  if (!india || typeof india !== "object") return india;
  var out = cleanSlice(india);
  if (india.quarters && typeof india.quarters === "object") {
    out.quarters = {};
    Object.keys(india.quarters).forEach(function (q) { out.quarters[q] = cleanSlice(india.quarters[q]); });
  }
  return out;
}

module.exports = { applyIndiaIncomeSwitches: applyIndiaIncomeSwitches };
